File: 431_data_engineering_with_go/msds431_week7/jump-start-mnist-iforest-main/r/getMNIST.R
# tested on R version 4.3.1
# documentation
# idx2r: Convert Files to and from IDX Format to Vectors, Matrices and Arrays
#     https://cran.r-project.org/web/packages/idx2r/index.html
# IDX is a format to store vector and arrays in binary format. 
# Reading IDX format is needed for instance to use the MNIST database 
# digits from http://yann.lecun.com/exdb/mnist/ provided by Yann LeCun.
# GitHub repository: https://github.com/edoffagne/idx2r

library(idx2r)
library(R.utils)

cat("\nRunning getMNIST.R . . . ")

# read in training images
fileName <- "../data/train-images-idx3-ubyte.gz"
images28x28 <- read_idx(gsub(pattern = "\\.gz", "", fileName))

cat("\nDimensions of images28x28:")
print(dim(images28x28))
# brute-force matrix transformation beginning with empty matrix
images784 <- matrix(, nrow = dim(images28x28)[1], 
                      ncol = dim(images28x28)[2]*dim(images28x28)[3], 
                      byrow = TRUE)
for (imageNo in 1:dim(images28x28)[1]) {
    image784Col <- 0
    for (imageRow in 1:dim(images28x28)[2]) {
        for (imageCol in 1:dim(images28x28)[3]) {
            image784Col <- image784Col + 1
            images784[imageNo, image784Col] <- images28x28[imageNo,imageRow,imageCol]
        }
    } 
}

cat("\nDimensions of images784")
print(dim(images784))

# save MNIST objects to R binary file
binaryFileName <- "MNIST_image_objects.RData"
save(images28x28, images784, file = binaryFileName)
cat("\nimages28x28 and images784 saved to", binaryFileName)

cat("\nFinished running getMNIST.R . . . \n")

File: 431_data_engineering_with_go/msds431_week7/jump-start-mnist-iforest-main/python/getMNIST.py
# tested under Python 3.9.13

import gzip
import numpy as np

# read in the training images
print("Running getMNIST.py . . . ")
print("\nReading MNIST training data\n")
with gzip.open('../data/train-images-idx3-ubyte.gz','r') as f:
   imageSize = 28
   numImages = 60000
   f.read(16)
   buffer = f.read(imageSize * imageSize * numImages)
   images = np.frombuffer(buffer, dtype=np.uint8).astype(np.float32)
   images28x28 = images.reshape(numImages, imageSize, imageSize, 1)
   images784 = images28x28.reshape(numImages, 28*28)
   
print("Shape of numpy array images28x28:", images28x28.shape)
print("First image from images28x28:")
print(images28x28[0,:,:,0])
print("Shape of numpy array images784:", images784.shape)
print("First image from images784:")
print(images784[0,:])

# read in training labels
with gzip.open('../data/train-labels-idx1-ubyte.gz','r') as f:
   numLabels = 60000
   f.read(8)
   buffer = f.read(numLabels)
   labels = np.frombuffer(buffer, dtype=np.uint8)

print("Shape of numpy array labels:", labels.shape)

print("Finished running getMNIST.py . . . ")


